src/Workspaces/VisualBasic/Portable/CodeGeneration/VisualBasicSyntaxGenerator.vb

```diff
@@ -840,78 +840,89 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeGeneration
-            Return Isolate(declaration, Function(decl) AsPrivateInterfaceImplementationInternal(decl, typeName))
+        Public Overrides Function AsPrivateInterfaceImplementation(declaration As SyntaxNode, interfaceTypeName As SyntaxNode, interfaceMemberName As String) As SyntaxNode
+            Return Isolate(declaration, Function(decl) AsPrivateInterfaceImplementationInternal(decl, interfaceTypeName, interfaceMemberName))
         End Function
 
-        Private Function AsPrivateInterfaceImplementationInternal(declaration As SyntaxNode, typeName As SyntaxNode) As SyntaxNode
-            Dim type = DirectCast(typeName, NameSyntax)
+        Private Function AsPrivateInterfaceImplementationInternal(declaration As SyntaxNode, interfaceTypeName As SyntaxNode, interfaceMemberName As String) As SyntaxNode
+            Dim type = DirectCast(interfaceTypeName, NameSyntax)
 
-            ' convert declaration statements to blocks
             declaration = WithBody(declaration, allowDefault:=False)
             declaration = WithAccessibility(declaration, Accessibility.Private)
 
-            Dim method = TryCast(declaration, MethodBlockSyntax)
-            If method IsNot Nothing Then
-                Dim interfaceMemberName = SyntaxFactory.IdentifierName(method.SubOrFunctionStatement.Identifier)
-
-                ' original method's name is used for interface member's name
-                Dim memberName = SyntaxFactory.IdentifierName(method.SubOrFunctionStatement.Identifier)
+            Dim memberName = If(interfaceMemberName IsNot Nothing, interfaceMemberName, GetMemberName(declaration))
+            declaration = WithName(declaration, GetNameAsIdentifier(interfaceTypeName) & "_" & memberName)
+            declaration = WithImplementsClause(declaration, SyntaxFactory.ImplementsClause(SyntaxFactory.QualifiedName(type, SyntaxFactory.IdentifierName(memberName))))
 
-                ' change actual method name to hide it
-                method = method.WithSubOrFunctionStatement(
-                    method.SubOrFunctionStatement.WithIdentifier(SyntaxFactory.Identifier(GetNameAsIdentifier(typeName) & "_" & GetNameAsIdentifier(memberName))))
+            Return declaration
+        End Function
 
-                ' add implements clause
-                Return method.WithSubOrFunctionStatement(
-                    method.SubOrFunctionStatement.WithImplementsClause(
-                        SyntaxFactory.ImplementsClause(SyntaxFactory.QualifiedName(type, interfaceMemberName))))
+        Private Function GetMemberName(declaration As SyntaxNode) As String
+            If GetImplementsClause(declaration) Is Nothing Then
+                Return GetName(declaration)
+            Else
+                Dim fullName = GetName(declaration)
+                If fullName IsNot Nothing Then
+                    Dim last = fullName.Split({"_"c}, StringSplitOptions.RemoveEmptyEntries).LastOrDefault()
```

Comment:
Do we need this heuristic? Given that `GetImplementsClause(declaration)` is non-null, can't we fetch the implemented member name from the implements clause?